src/vmm/src/acpi/x86_64.rs

```diff
@@ -0,0 +1,53 @@
+// Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::mem::size_of;
+
+use acpi_tables::fadt::{
+    IAPC_BOOT_ARG_FLAGS_MSI_NOT_PRESENT, IAPC_BOOT_ARG_FLAGS_PCI_ASPM,
+    IAPC_BOOT_ARG_FLAGS_VGA_NOT_PRESENT,
+};
+use acpi_tables::madt::{IoAPIC, LocalAPIC};
+use acpi_tables::{Aml, Fadt};
+use vm_memory::GuestAddress;
+use zerocopy::AsBytes;
+
+use crate::arch::x86_64::layout;
+use crate::Vmm;
+
+impl Vmm {
+    pub(crate) fn setup_interrupt_controllers(&self, nr_vcpus: u8) -> Vec<u8> {
+        let mut ic =
+            Vec::with_capacity(size_of::<IoAPIC>() + (nr_vcpus as usize) * size_of::<LocalAPIC>());
+
+        ic.extend_from_slice(IoAPIC::new(0, layout::IOAPIC_ADDR).as_bytes());
+        for i in 0..nr_vcpus {
+            ic.extend_from_slice(LocalAPIC::new(i).as_bytes());
+        }
+        ic
+    }
+
+    pub(crate) fn setup_arch_fadt(&self, fadt: &mut Fadt) {
+        // Let the guest kernel know that there is not VGA hardware present
+        // neither do we support ASPM, or MSI type of interrupts.
+        // More info here:
+        // https://uefi.org/specs/ACPI/6.5/05_ACPI_Software_Programming_Model.html?highlight=0a06#ia-pc-boot-architecture-flags
+        fadt.setup_iapc_flags(
+            1 << IAPC_BOOT_ARG_FLAGS_VGA_NOT_PRESENT
+                | 1 << IAPC_BOOT_ARG_FLAGS_PCI_ASPM
+                | 1 << IAPC_BOOT_ARG_FLAGS_MSI_NOT_PRESENT,
+        );
+    }
+
+    pub(crate) fn setup_arch_dsdt(&self, dsdt_data: &mut Vec<u8>) {
+        self.pio_device_manager.append_aml_bytes(dsdt_data);
+    }
+
+    pub(crate) fn get_apic_addr(&self) -> u32 {
+        layout::APIC_ADDR
+    }
+
+    pub(crate) fn rsdp_addr(&self) -> GuestAddress {
+        GuestAddress(layout::RSDP_ADDR)
+    }
```

Comment:
This again has to do with my preference for having architecture specific code in architecture specific modules/files. It is true that I can re-export from `vmm::arch`, so that I can just use `vmm::arch::RSDP_ADDR`, but atm we don't have Aarch64 support (so the `vmm::arch::aarch64::layout::RSDP_ADDR` doesn't yet exist), so I chose to hide this behind a function in the architecture specific module.